nixos/modules/services/hardware/nvidia-container-toolkit.nix

```diff
@@ -0,0 +1,84 @@
+            ${pkgs.jq}/bin/jq \
+               '.containerEdits.mounts[.containerEdits.mounts | length] |= . + { "hostPath": "/run/current-system/sw/bin/nvidia-cuda-mps-server", "containerPath": "/usr/bin/nvidia-cuda-mps-server", "options": ["ro", "nosuid", "nodev", "bind"] }' | \
+            ${pkgs.jq}/bin/jq \
+               '.containerEdits.mounts[.containerEdits.mounts | length] |= . + { "hostPath": "/run/current-system/sw/bin/nvidia-debugdump", "containerPath": "/usr/bin/nvidia-debugdump", "options": ["ro", "nosuid", "nodev", "bind"] }' | \
+            ${pkgs.jq}/bin/jq \
+               '.containerEdits.mounts[.containerEdits.mounts | length] |= . + { "hostPath": "/run/current-system/sw/bin/nvidia-powerd", "containerPath": "/usr/bin/nvidia-powerd", "options": ["ro", "nosuid", "nodev", "bind"] }' | \
+            ${pkgs.jq}/bin/jq \
+               '.containerEdits.mounts[.containerEdits.mounts | length] |= . + { "hostPath": "/run/current-system/sw/bin/nvidia-smi", "containerPath": "/usr/bin/nvidia-smi", "options": ["ro", "nosuid", "nodev", "bind"] }' | \
+            ${pkgs.jq}/bin/jq \
+               '.containerEdits.mounts[.containerEdits.mounts | length] |= . + { "hostPath": "${pkgs.glibc}/lib", "containerPath": "${pkgs.glibc}/lib", "options": ["ro", "nosuid", "nodev", "bind"] }' | \
+            ${pkgs.jq}/bin/jq \
+               '.containerEdits.mounts[.containerEdits.mounts | length] |= . + { "hostPath": "${config.hardware.nvidia.package}/lib", "containerPath": "${config.hardware.nvidia.package}/lib", "options": ["ro", "nosuid", "nodev", "bind"] }' | \
+            ${pkgs.jq}/bin/jq \
+               '.containerEdits.mounts[.containerEdits.mounts | length] |= . + { "hostPath": "${pkgs.nvidia-container-toolkit}/bin/nvidia-ctk", "containerPath": "/usr/bin/nvidia-ctk", "options": ["ro", "nosuid", "nodev", "bind"] }' > ${generatedCdiConfigPlacement}
+        ''; in "${script}/bin/nvidia-cdi-generator";
+      };
+    };
+
+    # This cleanup code is relevant in case the user is trying different configurations back and forth.
+    # According to the CDI spec [1], JSON files under `/var/run/cdi` (dynamic dir) have precedence over
+    # those at `/etc/cdi` (static dir). The latter ones are configured through `virtualisation.containers.cdi`.
+    # If the user is changing one and another back and forth trying the best setup for them, and we don't
+    # clean up the generated files, they might not see their changes applied to `/etc/cdi` due to our lack
+    # of cleanup at `/var/run/cdi`. If nvidia-container-toolkit is not enabled, and the generated file exists,
+    # clean it up.
+    #
+    # [1] https://github.com/cncf-tags/container-device-interface/blob/7b540ad63013eafbb6779fd06e3235653528d4ae/pkg/cdi/spec-dirs.go#L26-L31
+    systemd.services.nvidia-container-toolkit-cleanup = lib.mkIf (!config.hardware.nvidia-container-toolkit.enable) {
```

Comment:
It's not ideal that this unit will be present on all systems that don't have the nvidia-container-toolkit enabled.

If you have a better idea to handle this case it would be really great.

Reply:
Yes, you just wrap the entire `config` into `mkIf`

Reply:
Hm, so there are two systemd services:

1. `systemd.services.nvidia-container-toolkit` is enabled if `config.hardware.nvidia-container-toolkit.enable`. This systemd unit generates the `/var/run/cdi/nvidia-container-toolkit.json` based on the detected hardware.

2. `systemd.services.nvidia-container-toolkit-cleanup` is enabled if `!config.hardware.nvidia-container-toolkit.enable` (note the negation). This systemd unit removes `/var/run/cdi/nvidia-container-toolkit.json` if it exists.

The rationale behind `2.` is that if the user is trying different settings, we are cleaning up behind us.

Reply:
(semi-automatic reply, I may not have understood the previous message)

Does something like the following work?

```
config = mkMerge [
  (mkIf cfg.a.enable { ... })
  (mkIf cfg.b.enable { ... })
]
```

Reply:
> systemd.services.nvidia-container-toolkit-cleanup

Is there a chance we could everything to `systemd.exec`'s `RuntimeDirecty` and not implement the `cleanup`?

Reply:
No, not really. `systemd.services.nvidia-container-toolkit-cleanup` will always be applied if `!config.hardware.nvidia-container-toolkit.enable`.

This is my main concern with the solution that I'm proposing and what I'd like to improve.

Reply:
> Is there a chance we could everything to systemd.exec's RuntimeDirecty and not implement the cleanup?

I have updated the PR to use RuntimeDirectory. However, this is tangential. The main problem is, suppose a user does the following:

1. Sets `hardware.nvidia-container-toolkit.enable = true`. Switches to this configuration.
    1. Nix will create `/run/cdi/nvidia-container-toolkit.json`.
1. Sets `hardware.nvidia-container-toolkit.enable = false`. Switches to this configuration.
    1. We need to clean `/run/cdi/nvidia-container-toolkit.json` here.
    
If we don't clean up, there are chances that the user sets `virtualisation.containers.cdi`, but this is not applied because container runtimes are prioritizing the leftovers at `/run/cdi/nvidia-container-toolkit.json`.

Reply:
Once the user stops `systemd.services.nvidia-container-toolkit` (not sure that happens automatically when on `switch` but it might and eventually does anyway) the directory is cleaned by systemd

Reply:
We do not want to enable an nvidia-specific ("cleanup") service for non-nvidia users

Reply:
No, it's not working, just checked :(

Reply:
> We do not want to enable an nvidia-specific ("cleanup") service for non-nvidia users

I know we don't want to enable it, but it's the only way I saw how to fix https://github.com/NixOS/nixpkgs/pull/284507#discussion_r1477436036.

Reply:
Removed the cleanup script, finally :)

`Type = "oneshot"`, `RemainAfterExit = true` and cleaning up in `ExecStop` did the trick.